Adding HttpWebRequest to HttpClient Migration Guide (#42242)
dotnet/docs: docs/fundamentals/networking/http/httpclient-migrate-from-httpwebrequest.md

HttpWebRequest request = WebRequest.CreateHttp(uri);
using WebResponse response = await request.GetResponseAsync();
```

#### Simple GET Request Using <xref:System.Net.Http.HttpClient>

Here's an example of how the code might look:

```c#
using System.Net.Http;

using HttpClient client = new();
using HttpResponseMessage message = await client.GetAsync(uri);
```

#### Simple POST Request Using <xref:System.Net.HttpWebRequest>

Here's an example of how the code might look:

```c#
using System.Net;

HttpWebRequest request = WebRequest.CreateHttp(uri);
request.Method = "POST";
request.ContentType = "text/plain";
await using Stream stream = await request.GetRequestStreamAsync();
await stream.WriteAsync("Hello World!"u8.ToArray());
using WebResponse response = await request.GetResponseAsync();
Memory<byte> buffer = new byte[1024];
await using Stream responseStream = response.GetResponseStream();
```

#### Simple POST Request Using <xref:System.Net.Http.HttpClient>

Here's an example of how the code might look:

```c#
using System.Net.Http;

using HttpClient client = new();
using HttpResponseMessage responseMessage = await client.PostAsync(uri, new StringContent("Hello World!"));
```

> ## 1:1 API Mapping Table

Reviewer: it's not really 1:1 if some members don't have an equivalent API. Maybe something like this?

```suggestion
## Migrating ServicePointManager usage
```

### <xref:System.Net.ServicePointManager> Properties Mapping

| <xref:System.Net.ServicePointManager> Old API | New API | Notes |
|---------|----------------------|-------|
| `CheckCertificateRevocationList` | <xref:System.Net.Http.SocketsHttpHandler.SslOptions>.CertificateRevocationCheckMode | See TODO: for usage |
| `DefaultConnectionLimit` | <xref:System.Net.Http.SocketsHttpHandler.MaxConnectionsPerServer> | TODO |
| `DnsRefreshTimeout` | No equivalent API | Custom implementation needed |
| `EnableDnsRoundRobin` | No equivalent API | Custom implementation needed |
| `EncryptionPolicy` | No equivalent API | TODO |
| `Expect100Continue` | <xref:System.Net.Http.Headers.HttpRequestHeaders.ExpectContinue> | TODO |
| `MaxServicePointIdleTime` | <xref:System.Net.Http.SocketsHttpHandler.PooledConnectionIdleTimeout> | TODO |
| `MaxServicePoints` | No equivalent API | TODO |

> | `ReusePort` | No direct equivalent API | Can be done over <xref:System.Net.Http.SocketsHttpHandler.ConnectCallback>. |

Reviewer: "Can be done over ..." wording does not feel (stylistically) right, maybe just "See ConnectCallback" and add a paragraph with something like "Some ServicePointManager members don't have their equivalent on HttpClient, but connection customization can be performed on the underlying Socket instance by utilizing the ConnectCallback" before the tables?

| `SecurityProtocol` | <xref:System.Net.Http.SocketsHttpHandler.SslOptions>.EnabledSslProtocols | TODO |
| `ServerCertificateValidationCallback` | <xref:System.Net.Http.SocketsHttpHandler.SslOptions>.RemoteCertificateValidationCallback | Signatures are different need a custom callback |
| `UseNagleAlgorithm` | No direct equivalent API | Can be done over <xref:System.Net.Http.SocketsHttpHandler.ConnectCallback>. |

### <xref:System.Net.ServicePointManager> Method Mapping

| <xref:System.Net.ServicePointManager> Old API | New API | Notes |
|---------|----------------------|-------|
| `FindServicePoint` | No equivalent API | TODO |
| `SetTcpKeepAlive` | No direct equivalent API | Can be done over <xref:System.Net.Http.SocketsHttpHandler.ConnectCallback>. |

### <xref:System.Net.ServicePoint> Properties Mapping

| <xref:System.Net.ServicePoint> Old API | New API | Notes |
|---------|----------------------|-------|

> | `Address` | `HttpRequestMessage.RequestUri` | This is request uri, so basically we can get this from `HttpRequestMessage`. |

Reviewer: I think the ", so basically" filler words should be avoided, same for use of first person plural (we can....)

author: liveans
ms.author: aaksoy
ms.date: 07/25/2024
helpviewer_keywords: 
  - "protocols, HTTP"
  - "migration, HTTP"
  - "HTTP"
  - "HttpWebRequest class, about HttpWebRequest class"
  - "application protocols, HTTP"
  - "HttpClient class, about HttpClient class"
  - "data requests, HTTP"
  - "Internet, HTTP"
---

# HttpWebRequest to HttpClient Migration Guide

This document aims to guide developers through the process of migrating from <xref:System.Net.HttpWebRequest>, <xref:System.Net.ServicePoint>, and <xref:System.Net.ServicePointManager> to <xref:System.Net.Http.HttpClient>. The migration is necessary due to the obsolescence of the older APIs and the numerous benefits offered by <xref:System.Net.Http.HttpClient>, including improved performance, better resource management, and a more modern and flexible API design. By following the steps outlined in this document, developers will be able to transition their codebases smoothly and take full advantage of the features provided by <xref:System.Net.Http.HttpClient>.

## Migrating from <xref:System.Net.HttpWebRequest> to <xref:System.Net.Http.HttpClient>

> The migration from <xref:System.Net.HttpWebRequest> to <xref:System.Net.Http.HttpClient> is essential due to the modern features and improved performance that <xref:System.Net.Http.HttpClient> offers. <xref:System.Net.Http.HttpClient> provides a more flexible and efficient way to make HTTP requests and handle responses. Here are the key steps and considerations for this migration:

Reviewer: This paragraph seems to only repeat what was in the previous one, feels unnecessarily verbose.

By enabling DNS Round Robin, you can ensure that your application distributes requests evenly across multiple servers, reducing the load on any single server and improving overall performance and reliability. This approach is particularly useful for high-traffic applications that require efficient load balancing.

To enable DNS Round Robin with HttpClient, you can use the ConnectCallback property to manually resolve the DNS entries and rotate through the IP addresses. Here's an example for `HttpWebRequest` and `HttpClient`:

**Old Code Using `HttpWebRequest`**:

```csharp
ServicePointManager.DnsRefreshTimeout = 60000;
ServicePointManager.EnableDnsRoundRobin = true;
HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
HttpWebResponse response = (HttpWebResponse)request.GetResponse();
```

**New Code Using `HttpClient`**:

```csharp
ConcurrentDictionary<string, DnsCache> dnsCache = new(StringComparer.OrdinalIgnoreCase);

TimeSpan dnsRefreshTimeout = TimeSpan.FromMinutes(1);
var handler = new SocketsHttpHandler
{
    ConnectCallback = async (context, cancellationToken) =>
    {
        string host = context.DnsEndPoint.Host;
        
        DnsCache? cacheItem;

>         if (!dnsCache.TryGetValue(host, out cacheItem) || DateTime.Now > cacheItem.CacheExpireTime)

Reviewer 2: Don't use `DateTime.Now`, it's affected by daylight saving changes!

You can use `Environment.TickCount64` instead.

Reviewer 3: Shameless plug since this is quite a bit of code for someone to copy-paste: https://github.com/MihaZupan/DnsRoundRobin

        {
            dnsCache.TryRemove(host, out _);
            cacheItem = new DnsCache(await Dns.GetHostAddressesAsync(host, cancellationToken), DateTime.Now.Add(dnsRefreshTimeout), 0);

>             dnsCache.TryAdd(host, cacheItem);

Reviewer 2: `ConnectCallback` doesn't guarantee thread safety. This will work, but it might unnecessarily issue multiple DNS requests for the same host, while also unintentionally removing freshly inserted results.

        }

        IPAddress connectAddress;
        // If there is only one address, we don't need to do anything special
        if (cacheItem.Addresses.Length == 1)
        {
            connectAddress = cacheItem.Addresses[0];
        }
        else
        {
            int index = cacheItem.Index;
            connectAddress = cacheItem.Addresses[index];

>             cacheItem.IncreaseIndex();

Reviewer 2: Once again, not thread-safe.

            NoDelay = true
        };
        try
        {
            await socket.ConnectAsync(connectAddress, context.DnsEndPoint.Port, cancellationToken);
            return new NetworkStream(socket, ownsSocket: true);
        }
        catch
        {
            socket.Dispose();
            throw;
        }
    }
};
var client = new HttpClient(handler);
var response = await client.GetAsync(uri);

class DnsCache(IPAddress[] addresses, DateTime cacheExpireTime, int index)
{
    public IPAddress[] Addresses { get; } = addresses;
    public DateTime CacheExpireTime { get; } = cacheExpireTime;
    public int Index { get; private set; } = index;

    public void IncreaseIndex()
    {
        Index = (Index + 1) % Addresses.Length;
    }
}
```

### Example: Setting SocketsHttpHandler Properties

TODO:

### Example: Change ImpersonationLevel

TODO:

## Usage of Certificate and TLS Related Properties in HttpClient

When working with `HttpClient`, you may need to handle client certificates for various purposes, such as custom validation of server certificates or fetching the server certificate. `HttpClient` provides several properties and options to manage certificates effectively.

> ### Example: Enabling CRL Check with SocketsHttpHandler

Reviewer 2: NIT: I don't like using abbreviations before they are introduced, so I'd opt to something like:

Check certificate revocation list with `SocketsHttpHandler`

> The `CheckCertificateRevocationList` property in `SocketsHttpHandler` allows developers to enable or disable the check for certificate revocation lists (CRL) during SSL/TLS handshake. Enabling this property ensures that the client verifies whether the server's certificate has been revoked, enhancing the security of the connection.

Reviewer 2: NIT: It's not a property of the handler, but of the `SslOptions` ==>
```suggestion
The `CheckCertificateRevocationList` property in `SocketsHttpHandler.SslOptions` allows developers to enable or disable the check for certificate revocation lists (CRL) during SSL/TLS handshake. Enabling this property ensures that the client verifies whether the server's certificate has been revoked, enhancing the security of the connection.
```

**Old Code Using `HttpWebRequest`**:

```csharp
ServicePointManager.CheckCertificateRevocationList = true;
HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
HttpWebResponse response = (HttpWebResponse)request.GetResponse();
```

**New Code Using `HttpClient`**:

```csharp
bool checkCertificateRevocationList = true;
var handler = new SocketsHttpHandler
{
    SslOptions =
    {
        CertificateRevocationCheckMode = checkCertificateRevocationList ? X509RevocationMode.Online : X509RevocationMode.NoCheck,
    }
};
var client = new HttpClient(handler);
var response = await client.GetAsync(uri);
```

### Example: Fetch Certificate

> To fetch the certificate from the RemoteCertificateValidationCallback in HttpClient, you can use the ServerCertificateCustomValidationCallback property of HttpClientHandler or SocketsHttpHandler. This callback allows you to inspect the server's certificate during the SSL/TLS handshake.

Reviewer 2: It's actually different callback with different shape, so either I'd only mention the `SocketsHttpHandler.SslOptions` one, or split the new code into 2 examples.

NITs:
```suggestion
To fetch the certificate from the `RemoteCertificateValidationCallback` in `HttpClient`, you can use the `ServerCertificateCustomValidationCallback` property of `HttpClientHandler` or `SocketsHttpHandler.SslOptions`. This callback allows you to inspect the server's certificate during the SSL/TLS handshake.
```

### Example: Binding IP Address to Socket

In the old approach using `HttpWebRequest`, you might have used custom logic to bind a specific IP address to a socket. Here's how you can achieve similar functionality using `HttpClient` and `ConnectCallback`:

**Old Code Using `HttpWebRequest`**:

```csharp
HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
request.ServicePoint.BindIPEndPointDelegate = (servicePoint, remoteEndPoint, retryCount) =>
{
    // Bind to a specific IP address
    IPAddress localAddress = IPAddress.Parse("192.168.1.100");
    return new IPEndPoint(localAddress, 0);
};
using HttpWebResponse response = (HttpWebResponse)request.GetResponse();
```

**New Code Using `HttpClient` and `ConnectCallback`**:

```csharp
var handler = new SocketsHttpHandler
{
    ConnectCallback = async (context, cancellationToken) =>
    {
        // Bind to a specific IP address
        IPAddress localAddress = IPAddress.Parse("192.168.1.100");

>         var socket = new Socket(localAddress.AddressFamily, SocketType.Stream, ProtocolType.Tcp);

Reviewer 2: You need a try-catch block to eventually dispose the socket if connect fails, see: https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-8.0#examples
Applies to other examples bellow as well.

```

**Simple GET Request Using <xref:System.Net.Http.HttpClient>**

Here's an example of how the code might look:

```c#
using System.Net.Http;

HttpClient client = new();
using HttpResponseMessage message = await client.GetAsync(uri);
```

**Simple POST Request Using <xref:System.Net.HttpWebRequest>**

Here's an example of how the code might look:

```c#
using System.Net;

HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
request.Method = "POST";
request.ContentType = "text/plain";
await using Stream stream = await request.GetRequestStreamAsync();
await stream.WriteAsync("Hello World!"u8.ToArray());
using WebResponse response = await request.GetResponseAsync();
Memory<byte> buffer = new byte[1024];
await using Stream responseStream = response.GetResponseStream();
```

**Simple POST Request Using <xref:System.Net.Http.HttpClient>**

Here's an example of how the code might look:

```c#
using System.Net.Http;

HttpClient client = new();
using HttpResponseMessage responseMessage = await client.PostAsync(uri, new StringContent("Hello World!"));
```

## HttpWebRequest to HttpClient, SocketsHttpHandler Migration Guide

> | <xref:System.Net.HttpWebRequest> Old API | New API | Notes |

Reviewer 3: Nit: wording is inconsistent between "No direct equivalent API" and "No equivalent API"

Author: It's not inconsistent actually, `No direct equivalent API` means there is a workaround/different shaped usable API, `No equivalent API` means there is no workaround/usable API

| `AutomaticDecompression` | <xref:System.Net.Http.SocketsHttpHandler.AutomaticDecompression> | See: [Example: Setting SocketsHttpHandler Properties](#example-setting-socketshttphandler-properties). |
| `CachePolicy` | No direct equivalent API | See: [Example: Apply CachePolicy Headers](#example-apply-cachepolicy-headers). |
| `ClientCertificates` | <xref:System.Net.Http.SocketsHttpHandler.SslOptions>.<xref:System.Net.Security.SslClientAuthenticationOptions.ClientCertificates> | See: [Usage of Certificate Related Properties in HttpClient](#usage-of-certificate-and-tls-related-properties-in-httpclient). |
| `Connection` | <xref:System.Net.Http.Headers.HttpRequestHeaders.Connection> | See: [Example: Set Request Headers](#example-set-common-request-headers). |
| `ConnectionGroupName` | No equivalent API | No workaround |
| `ContentLength` | <xref:System.Net.Http.Headers.HttpContentHeaders.ContentLength> | See: [Example: Set Content Headers](#example-set-content-headers). |
| `ContentType` | <xref:System.Net.Http.Headers.HttpContentHeaders.ContentType> | See: [Example: Set Content Headers](#example-set-content-headers). |

> | `ContinueDelegate` | No equivalent API | There is no workaround. |

Reviewer 3: ```suggestion
| `ContinueDelegate` | No equivalent API | No workaround. |
```

| `Credentials` | <xref:System.Net.Http.SocketsHttpHandler.Credentials> | See: [Example: Setting SocketsHttpHandler Properties](#example-setting-socketshttphandler-properties). |
| `Date` | <xref:System.Net.Http.Headers.HttpRequestHeaders.Date> | See: [Example: Set Request Headers](#example-set-common-request-headers). |
| `DefaultCachePolicy` | No direct equivalent API | See: [Example: Apply CachePolicy Headers](#example-apply-cachepolicy-headers). |
| `DefaultMaximumErrorResponseLength` | No direct equivalent API | See: [Example: Set MaximumErrorResponseLength in HttpClient](#example-set-maximumerrorresponselength-in-httpclient). |
| `DefaultMaximumResponseHeadersLength` | No equivalent API | <xref:System.Net.Http.SocketsHttpHandler.MaxResponseHeadersLength> can be used instead. |
| `DefaultWebProxy` | No equivalent API | <xref:System.Net.Http.SocketsHttpHandler.Proxy> can be used instead. |
| `Expect` | <xref:System.Net.Http.Headers.HttpRequestHeaders.Expect> | See: [Example: Set Request Headers](#example-set-common-request-headers). |

> | `HaveResponse` | No equivalent API | No workaround |

Reviewer 3: ```suggestion
| `HaveResponse` | No equivalent API | Implied by having an `HttpResponseMessage` instance. |
```

> In modern .NET, the default values for the `UseNagleAlgorithm` and `Expect100Continue` properties are set to `false`. These values were `true` by default in .NET Framework.

### <xref:System.Net.ServicePointManager> Method Mapping

| <xref:System.Net.ServicePointManager> Old API | New API | Notes |
|---------|----------------------|-------|
| `FindServicePoint` | No equivalent API | No workaround |
| `SetTcpKeepAlive` | No direct equivalent API | See: [Usage of SocketsHttpHandler and ConnectCallback](#usage-of-socketshttphandler-and-connectcallback). |

### <xref:System.Net.ServicePoint> Properties Mapping

| <xref:System.Net.ServicePoint> Old API | New API | Notes |
|---------|----------------------|-------|
| `Address` | `HttpRequestMessage.RequestUri` | This is request uri, this information can be found under `HttpRequestMessage`. |
| `BindIPEndPointDelegate` | No direct equivalent API | See: [Usage of SocketsHttpHandler and ConnectCallback](#usage-of-socketshttphandler-and-connectcallback). |
| `Certificate` | No direct equivalent API | This information can be fetched from `RemoteCertificateValidationCallback`. See: [Example: Fetch Certificate](#example-fetch-certificate) |

> | `ClientCertificate` | No equivalent API | No workaround |

Reviewer 3: Nit: this can link to your example of configuring the client cert

| `ConnectionLeaseTimeout` | `SocketsHttpHandler.PooledConnectionLifetime` | Equivalent setting in <xref:System.Net.Http.HttpClient> |
| `ConnectionLimit` | <xref:System.Net.Http.SocketsHttpHandler.MaxConnectionsPerServer> | See: [Example: Setting SocketsHttpHandler Properties](#example-setting-socketshttphandler-properties). |
| `ConnectionName` | No equivalent API | No workaround |

> | `CurrentConnections` | No equivalent API | No workaround |

Reviewer 3: ```suggestion
| `CurrentConnections` | No equivalent API | See [Networking telemetry in .NET docs](https://learn.microsoft.com/dotnet/fundamentals/networking/telemetry/overview). |
```

### Example: Binding IP Address to Socket

In the old approach using `HttpWebRequest`, you might have used custom logic to bind a specific IP address to a socket. Here's how you can achieve similar functionality using `HttpClient` and `ConnectCallback`:

**Old Code Using `HttpWebRequest`**:

```csharp
HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
request.ServicePoint.BindIPEndPointDelegate = (servicePoint, remoteEndPoint, retryCount) =>
{
    // Bind to a specific IP address
    IPAddress localAddress = IPAddress.Parse("192.168.1.100");
    return new IPEndPoint(localAddress, 0);
};
using HttpWebResponse response = (HttpWebResponse)request.GetResponse();
```

**New Code Using `HttpClient` and `ConnectCallback`**:

```csharp
var handler = new SocketsHttpHandler
{
    ConnectCallback = async (context, cancellationToken) =>
    {
        // Bind to a specific IP address
        IPAddress localAddress = IPAddress.Parse("192.168.1.100");
        var socket = new Socket(localAddress.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
        try
        {
            socket.Bind(new IPEndPoint(localAddress, 0));

>             await socket.ConnectAsync(context.DnsEndPoint, cancellationToken);
>         }
>         catch
>         {
>             socket.Dispose();
>             throw;
>         }
>         return new NetworkStream(socket, ownsSocket: true);

Reviewer 3: ```suggestion
            await socket.ConnectAsync(context.DnsEndPoint, cancellationToken);
            return new NetworkStream(socket, ownsSocket: true);
        }
        catch
        {
            socket.Dispose();
            throw;
        }
```

HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
request.ServicePoint.SetTcpKeepAlive(true, 60000, 1000);
request.ServicePoint.ReceiveBufferSize = 8192;
request.ServicePoint.UseNagleAlgorithm = false;
using HttpWebResponse response = (HttpWebResponse)request.GetResponse();
```

**New Code Using `HttpClient` and `ConnectCallback`**:

```csharp
var handler = new SocketsHttpHandler
{
    ConnectCallback = async (context, cancellationToken) =>
    {
        var socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
        try
        {
            // Setting TCP Keep Alive
            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, 60);
            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, 1);

            // Setting ReceiveBufferSize
            socket.ReceiveBufferSize = 8192;

            // Enabling ReusePort
            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseUnicastPort, true);

            // Disabling Nagle Algorithm
            socket.NoDelay = true;

>             await socket.ConnectAsync(context.DnsEndPoint, cancellationToken);
>         }
>         catch
>         {
>             socket.Dispose();
>             throw;
>         }
>         return new NetworkStream(socket, ownsSocket: true);

Reviewer 3: ```suggestion
            await socket.ConnectAsync(context.DnsEndPoint, cancellationToken);
            return new NetworkStream(socket, ownsSocket: true);
        }
        catch
        {
            socket.Dispose();
            throw;
        }
```

To enable DNS Round Robin with HttpClient, you can use the ConnectCallback property to manually resolve the DNS entries and rotate through the IP addresses. Here's an example for `HttpWebRequest` and `HttpClient`:

**Old Code Using `HttpWebRequest`**:

```csharp
ServicePointManager.DnsRefreshTimeout = 60000;
ServicePointManager.EnableDnsRoundRobin = true;
HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
using HttpWebResponse response = (HttpWebResponse)request.GetResponse();
```

In the older `HttpWebRequest` API, enabling DNS Round Robin was straightforward due to its built-in support for this feature. However, the newer `HttpClient` API does not provide the same built-in functionality. Despite this, you can achieve similar behavior by implementing a `DnsRoundRobinConnector` that manually rotates through the IP addresses returned by DNS resolution.

**New Code Using `HttpClient`**:

> :::code source="../snippets/httpclient/DnsRoundRobin.cs" id="DnsRoundRobinConnector":::

Reviewer 3: I still find it odd we've decided to copy an entire project into a sample rather than linking to the NuGet package/source

Author: We might also reference to NuGet package, I guess and keep the sample here.

**Setting Common Default Request Headers in HttpClient**

```csharp
using System.Net.Http.Headers;

var client = new HttpClient();
client.DefaultRequestHeaders.Authorization =  new AuthenticationHeaderValue("Bearer", "token");
```

**Setting Common Request Headers in HttpRequestMessage**

```csharp
using System.Net.Http.Headers;

var request = new HttpRequestMessage(HttpMethod.Get, uri);
request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", "token");
```

### Example: Set Content Headers

Content headers are used to provide additional information about the body of an HTTP request or response. When working with `HttpClient` in .NET, you can set content headers to specify the media type, encoding, and other metadata related to the content being sent or received. Properly configuring content headers ensures that the server and client can correctly interpret and process the content.

```csharp
var client = new HttpClient();
var request = new HttpRequestMessage(HttpMethod.Post, uri);

// Create the content and set the content headers
var jsonData = "{\"key\":\"value\"}";
var content = new StringContent(jsonData, Encoding.UTF8, "application/json");

> // This is also corresponding example to set `MediaType` in `HttpWebRequest`.
> content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
> content.Headers.ContentEncoding.Add("utf-8");
> // WARNING: This is not needed in most of the cases, because most `HttpContent` implementations will calculate this field automatically.
> content.Headers.ContentLength = jsonData.Length;

Reviewer 3: ```suggestion

// The following headers are set automatically by `StringContent`. If you wish to override their values, you can do it like so:
// content.Headers.ContentType = new MediaTypeHeaderValue("application/json; charset=utf-8");
// content.Headers.ContentLength = Encoding.UTF8.GetByteCount(jsonData);
```
`StringContent` already sets the content-type for you, also accounting for the encoding used.
Despite the name, `ContentEncoding` isn't used to communicate the type of text encoding used.

// Assign the content to the request
request.Content = content;

using var response = await client.SendAsync(request);
```

### Example: Set MaximumErrorResponseLength in HttpClient

The `MaximumErrorResponseLength` usage allows developers to specify the maximum length of the error response content that the handler will buffer. This is useful for controlling the amount of data that is read and stored in memory when an error response is received from the server. By using this technique, you can prevent excessive memory usage and improve the performance of your application when handling large error responses.

There are couple of ways to do that, we'll examine `TruncatedReadStream` technique on this example:

:::code source="../snippets/httpclient/Program.TruncatedReadStream.cs" id="TruncatedReadStream":::

And usage example of `TruncatedReadStream`:

:::code source="../snippets/httpclient/Program.TruncatedReadStream.cs" id="TruncatedReadStreamUsage":::

> ### Example: Apply CachePolicy Headers

Reviewer 3: We may want to delete this section/example as it's a bit misleading.
With WebRequest, you would set these settings to control how the framework itself handles the caching of responses for you, not just to set headers.
HttpClient does no such caching, so just setting the headers isn't achieving anything on its own.

Author: We might add another explanation about how this actually should work normally, leave this as-is, so people can get the idea about how these headers can be set.

When migrating from `HttpWebRequest` to `HttpClient`, it's important to correctly handle cache-related headers such as `pragma` and `cache-control`. These headers control how responses are cached and retrieved, ensuring that your application behaves as expected in terms of performance and data freshness.

In `HttpWebRequest`, you might have used the `CachePolicy` property to set these headers. However, in `HttpClient`, you need to manually set these headers on the request.

**Old Code Using `HttpWebRequest`**:

```csharp

> HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);

Reviewer 3: ```suggestion
HttpWebRequest request = WebRequest.CreateHttp(uri);
```

## Usage of Buffering Properties

When migrating from HttpWebRequest to `HttpClient`, it's important to understand the differences in how these two APIs handle buffering.

**Old Code Using `HttpWebRequest`**:

In `HttpWebRequest`, you have direct control over buffering properties through the `AllowWriteStreamBuffering` and `AllowReadStreamBuffering` properties. These properties enable or disable buffering of data sent to and received from the server.

```csharp
HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);
request.AllowReadStreamBuffering = true; // Default is `false`.
request.AllowWriteStreamBuffering = false; // Default is `true`.
```

**New Code Using `HttpClient`**:

In `HttpClient`, there are no direct equivalents to the `AllowWriteStreamBuffering` and `AllowReadStreamBuffering` properties.

However, you can achieve similar outcome of `disable write buffering` behavior by using the `StreamContent` class for streaming data. Here's an example:

```csharp
HttpClient client = new HttpClient();

HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, uri);
request.Content = new StreamContent(yourStream);

> request.Headers.TransferEncodingChunked = true;

Reviewer 3: ```suggestion
```

Whether chunked encoding is used or not is irrelevant to whether the request will be buffered or not.

HttpResponseMessage response = await client.SendAsync(request);
```

> In `HttpClient` read buffering is enabled by default. `Disabling read buffering` is much easier. Here's an example:
> 
> ```csharp
> HttpClient client = new HttpClient();
> 
> HttpResponseMessage response = await client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead);

Reviewer 3: ```suggestion
In `HttpClient` read buffering is enabled by default. To avoid it, you may specify the `HttpCompletionOption.ResponseHeadersRead` flag, or use the `GetStreamAsync` helper.

```csharp
HttpClient client = new HttpClient();

using HttpResponseMessage response = await client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead);
await using Stream responseStream = await response.Content.ReadAsStreamAsync();

// Or simply
await using Stream responseStream = await client.GetStreamAsync(uri);
```

---
title: Migrate from HttpWebRequest
description: Guidance document for migrating from HttpWebRequest to HttpClient.
author: liveans
ms.author: aaksoy
ms.date: 07/25/2024
helpviewer_keywords: 
  - "protocols, HTTP"
  - "migration, HTTP"
  - "HTTP"
  - "HttpWebRequest class, about HttpWebRequest class"
  - "application protocols, HTTP"
  - "HttpClient class, about HttpClient class"
  - "data requests, HTTP"
  - "Internet, HTTP"
---

# HttpWebRequest to HttpClient Migration Guide

> This document aims to guide developers through the process of migrating from <xref:System.Net.HttpWebRequest>, <xref:System.Net.ServicePoint>, and <xref:System.Net.ServicePointManager> to <xref:System.Net.Http.HttpClient>. The migration is necessary due to the obsolescence of the older APIs and the numerous benefits offered by <xref:System.Net.Http.HttpClient>, including improved performance, better resource management, and a more modern and flexible API design. By following the steps outlined in this document, developers will be able to transition their codebases smoothly and take full advantage of the features provided by <xref:System.Net.Http.HttpClient>.

Reviewer 3: While mentioning performance, should we mention that it's not just a "nice to have" bump in perf by migrating.
It's really that their existing WebRequest logic's perf is likely going to completely tank once they move to Core since WebRequest is just a minimal compat layer (e.g. no connection reuse in tons of cases).

Author: I was actually thinking to add a note or warning for this, to make sure that people who reads this doc will be aware of that `WebRequest` is compat layer on top of `HttpClient`.

| `SendChunked` | <xref:System.Net.Http.Headers.HttpRequestHeaders.TransferEncodingChunked> | See: [Example: Set Request Headers](#example-set-request-headers). |
| `ServerCertificateValidationCallback` | <xref:System.Net.Http.SocketsHttpHandler.SslOptions>.<xref:System.Net.Security.SslClientAuthenticationOptions.RemoteCertificateValidationCallback> | See: [Example: Setting SocketsHttpHandler Properties](#example-setting-socketshttphandler-properties). |
| `ServicePoint` | No equivalent API | `ServicePoint` is not part of `HttpClient`. |
| `SupportsCookieContainer` | No equivalent API | This is always `true` for `HttpClient`. |
| `Timeout` | <xref:System.Net.Http.HttpClient.Timeout> |  |
| `TransferEncoding` | <xref:System.Net.Http.Headers.HttpRequestHeaders.TransferEncoding> | See: [Example: Set Request Headers](#example-set-request-headers). |
| `UnsafeAuthenticatedConnectionSharing` | No equivalent API | No workaround |
| `UseDefaultCredentials` | No direct equivalent API | See: [Example: Setting SocketsHttpHandler Properties](#example-setting-socketshttphandler-properties). |

> | `UserAgent` | <xref:System.Net.Http.Headers.HttpRequestHeaders.UserAgent> | See: [Example: Set Request Headers](#example-set-request-headers). |

Reviewer 3: Nit: A bunch of these are headers names that have 1:1 mappings.
It might be easier to follow if those have a separate "For setting headers like `MediaType`, `Referer`, ... see [Example])".
On the other hand your list is in alphabetical order so maybe that's better 🤷‍♂️

Author: I'm not sure, I believe having these properties in alphabetical order would be more useful. Because it's much easier to think about something like this: I need to find X property under Y class, so I can check this Z list.

## Migrating ServicePoint(Manager) usage

> Developers should be aware that `ServicePointManager` is a static class, meaning that any changes made to its properties will have a global effect on all newly created `ServicePoint` objects within the application. For example, modifying a property like `ConnectionLimit` or `Expect100Continue` will impact every new ServicePoint instance.

Reviewer 3: I would add an extra warning here that modern `HttpClient` will completely ignore any settings they may have set on SPM, so they MUST switch over.

| `DnsRefreshTimeout` | No equivalent API | See: [Example: Enabling Dns Round Robin](#example-enabling-dns-round-robin). |
| `EnableDnsRoundRobin` | No equivalent API | See: [Example: Enabling Dns Round Robin](#example-enabling-dns-round-robin). |
| `EncryptionPolicy` | No equivalent API | TODO |
| `Expect100Continue` | <xref:System.Net.Http.Headers.HttpRequestHeaders.ExpectContinue> | TODO |
| `MaxServicePointIdleTime` | <xref:System.Net.Http.SocketsHttpHandler.PooledConnectionIdleTimeout> | TODO |
| `MaxServicePoints` | No equivalent API | `ServicePoint` is not part of `HttpClient`. |
| `ReusePort` | No direct equivalent API | See <xref:System.Net.Http.SocketsHttpHandler.ConnectCallback>. |
| `SecurityProtocol` | <xref:System.Net.Http.SocketsHttpHandler.SslOptions>.<xref:System.Net.Security.SslClientAuthenticationOptions.EnabledSslProtocols> | TODO |
| `ServerCertificateValidationCallback` | <xref:System.Net.Http.SocketsHttpHandler.SslOptions>.<xref:System.Net.Security.SslClientAuthenticationOptions.RemoteCertificateValidationCallback> | Both of them are <xref:System.Net.Security.RemoteCertificateValidationCallback> |

> | `UseNagleAlgorithm` | No direct equivalent API | See <xref:System.Net.Http.SocketsHttpHandler.ConnectCallback>. |

Reviewer 3: It might be worth calling out here that the defaults are different.

On Core, `Expect100Continue` is off by default.
On Core, `UseNagleAlgorithm` is off by default, so they likely don't need to touch `ConnectCallback`.

|---------|----------------------|-------|
| `CloseConnectionGroup` | No equivalent | No workaround |
| `SetTcpKeepAlive` | No direct equivalent API | See <xref:System.Net.Http.SocketsHttpHandler.ConnectCallback>. |

## Usage of HttpClient and HttpRequestMessage Properties

When working with HttpClient in .NET, you have access to a variety of properties that allow you to configure and customize HTTP requests and responses. Understanding these properties can help you make the most of HttpClient and ensure that your application communicates efficiently and securely with web services.

### Example: Usage of HttpRequestMessage properties

Here's an example of how to use HttpClient and HttpRequestMessage together:

```csharp
var client = new HttpClient();

var request = new HttpRequestMessage(HttpMethod.Post, "https://example.com"); // Method and RequestUri usage
var request = new HttpRequestMessage() // Alternative way to set RequestUri and Method
{
    RequestUri = new Uri("https://example.com"),
    Method = HttpMethod.Post
};
request.Headers.Add("Custom-Header", "value");
request.Content = new StringContent("somestring");

> var response = await client.SendAsync(request);

Reviewer 3: ```suggestion
using var response = await client.SendAsync(request);
```

Similar concern as https://github.com/dotnet/docs/pull/42242#discussion_r1753860667, we should try to dispose responses in all example code since people will just copy-paste them

Here's an example of how to configure SocketsHttpHandler with various properties and use it with HttpClient:

```c#
var cookieContainer = new CookieContainer();
cookieContainer.Add(new Cookie("cookieName", "cookieValue"));

var handler = new SocketsHttpHandler
{
    AllowAutoRedirect = true,
    AutomaticDecompression = DecompressionMethods.All,
    Expect100ContinueTimeout = TimeSpan.FromSeconds(1),
    CookieContainer = cookieContainer,
    Credentials = new NetworkCredential("user", "pass"),
    MaxAutomaticRedirections = 10,
    MaxResponseHeadersLength = 1,
    Proxy = new WebProxy("http://proxyserver:8080"), // Don't forget to set UseProxy
    UseProxy = true,
    SslOptions = new SslClientAuthenticationOptions
    {
        RemoteCertificateValidationCallback = (sender, certificate, chain, sslPolicyErrors) =>
        {
            // Custom validation logic
            return sslPolicyErrors == SslPolicyErrors.None;
        }
    },
};

var client = new HttpClient(handler);
var response = await client.GetAsync(uri);
```

### Example: Change ImpersonationLevel

> [!WARNING]

> > This example contains the use of reflection, which may affect the performance of your application and can be broken in future versions of .NET. Use reflection with caution and only when necessary.

Reviewer 3: I'd be more in favor of documenting something along the lines of
"this functionality is platform-specific and somewhat obsolete. If you _really_ need a workaround, see https://github.com/dotnet/runtime/blob/171f1a73a9f0fa77464995bcb893a59b9b98bc3d/src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L1678-L1684"
and deleting this example from the doc

Debug.Assert(settings != null);
FieldInfo? fi = Type.GetType("System.Net.Http.HttpConnectionSettings, System.Net.Http")?.GetField("_impersonationLevel", BindingFlags.NonPublic | BindingFlags.Instance);
fi?.SetValue(settings, TokenImpersonationLevel.Impersonation);

var client = new HttpClient(handler);
var response = await client.GetAsync(uri);
```

## Usage of Certificate and TLS Related Properties in HttpClient

When working with `HttpClient`, you may need to handle client certificates for various purposes, such as custom validation of server certificates or fetching the server certificate. `HttpClient` provides several properties and options to manage certificates effectively.

### Example: Enabling CRL Check with SocketsHttpHandler

The `CheckCertificateRevocationList` property in `SocketsHttpHandler` allows developers to enable or disable the check for certificate revocation lists (CRL) during SSL/TLS handshake. Enabling this property ensures that the client verifies whether the server's certificate has been revoked, enhancing the security of the connection.

**Old Code Using `HttpWebRequest`**:

```csharp
ServicePointManager.CheckCertificateRevocationList = true;

> HttpWebRequest request = (HttpWebRequest)WebRequest.Create(uri);

Reviewer 3: Nit: some places use `CreateHttp` instead of `Create`. Might as well be consistent, unless this was intentional?

client.DefaultRequestHeaders.Authorization =  new AuthenticationHeaderValue("Bearer", "token");
```

**Setting Common Request Headers in HttpRequestMessage**

```csharp
using System.Net.Http.Headers;

var request = new HttpRequestMessage(HttpMethod.Get, uri);
request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", "token");
```

### Example: Set Content Headers

Content headers are used to provide additional information about the body of an HTTP request or response. When working with `HttpClient` in .NET, you can set content headers to specify the media type, encoding, and other metadata related to the content being sent or received. Properly configuring content headers ensures that the server and client can correctly interpret and process the content.

```csharp
var client = new HttpClient();
var request = new HttpRequestMessage(HttpMethod.Post, uri);

// Create the content and set the content headers
var jsonData = "{\"key\":\"value\"}";
var content = new StringContent(jsonData, Encoding.UTF8, "application/json");
content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
content.Headers.ContentEncoding.Add("utf-8");

> content.Headers.ContentLength = jsonData.Length;

Reviewer 3: ```suggestion
```

I think we should avoid giving an example for this one.
It's unlikely that the user actually needs it (in this example, `StringContent` already calculates it for you), and it's easy to get wrong.

E.g. in this example, if you change `jsonData` to contain non-ASCII text, you'll break the content since the length will be wrong (number of UTF-16 chars instead of UTF-8 bytes).

Author: I think we can still keep this example here with extra warning like you said, but we should also add a note that they will not need this and Content itself will calculate `ContentLength` automatically based on the content data.

// Assign the content to the request
request.Content = content;

var response = await client.SendAsync(request);
```

### Example: Set MaximumErrorResponseLength in HttpClient

The `MaximumErrorResponseLength` usage allows developers to specify the maximum length of the error response content that the handler will buffer. This is useful for controlling the amount of data that is read and stored in memory when an error response is received from the server. By using this technique, you can prevent excessive memory usage and improve the performance of your application when handling large error responses.

There are couple of ways to do that, we'll examine `TruncatedReadStream` technique on this example:

```csharp
using System.Net;

int maxErrorResponseLength = 1 * 1024; // 1 KB

var client = new HttpClient();
var response = await client.GetAsync(uri);

if (response.Content is not null)
{
    var responseReadStream = response.Content.ReadAsStream();
    // If MaxErrorResponseLength is set and the response status code is an error code, then wrap the response stream in a TruncatedReadStream

>     if (maxErrorResponseLength >= 0 && response.StatusCode >= HttpStatusCode.BadRequest)

Reviewer 3: ```suggestion
    if (maxErrorResponseLength >= 0 && !response.IsSuccessStatusCode)
```
